stellargraph/core/graph.py

```diff
@@ -599,7 +625,51 @@ def info(self, show_attributes=True, sample=None):
         if self._graph is not None:
             return self._graph.info(show_attributes, sample)
 
-        raise NotImplementedError()
+        directed_str = "Directed" if self.is_directed() else "Undirected"
+        lines = [
+            f"{type(self)}: {directed_str} multigraph",
+            f" Nodes: {self.number_of_nodes()}, Edges: {self.number_of_edges()}",
+        ]
+
+        # Sample the nodes for our analysis
+        if sample:
+            all_nodes = list(self._graph.nodes)
+            snodes = random.sample(all_nodes, sample)
+        else:
+            snodes = None
+
+        gs = self.create_graph_schema(nodes=snodes)
+
+        def str_edge_type(et):
+            n1, rel, n2 = et
+            return f"{n1}-{rel}->{n2}"
+
+        lines.append("")
+        lines.append(" Node types:")
+
+        for nt in gs.node_types:
+            nodes = self.nodes_of_type(nt)
+            lines.append(f"  {nt}: [{len(nodes)}]")
+            edge_types = ", ".join(str_edge_type(et) for et in gs.schema[nt])
+            lines.append(f"    Edge types: {edge_types}")
+
+        lines.append("")
+        lines.append(" Edge types:")
+
+        all_type_ilocs = self._edge_type_iloc_triples(stacked=True)
```

Comment:
Good point :+1: My inclination would be to remove the sampling because I would guess the numpy-based version is significantly faster, removing the necessity of sampling. I will investigate this tomorrow.

Reply:
### Summary

- this new method *is* significantly faster
- it also scales much better (I think it's likely basically linear in the number of edges plus some factor for the number of edge type triples, for real graphs)
- the sampling doesn't seem to actually help the networkx implementation that much, even with a sample that's much smaller than the whole graph...

As such, I'd be fine just removing the sampling in the new code. Potentially, we can get even bigger wins with new code + sampling but maybe we can do that as future work?

### Details 

Times in milliseconds, `-` means I didn't want to wait for the measurement because it was likely to be slow:

First, for different number of node types and edge types, with 1k nodes, 20k edges:

| node types | edge types |  new |  old | old (`sample=20`) |
|-----------:|-----------:|-----:|-----:|------------------:|
|          1 |          1 | 37.3 |  244 |               132 |
|          1 |          2 | 36.5 |  366 |               244 |
|          2 |          1 | 45.1 |  610 |               475 |
|          2 |          2 | 45.0 | 1040 |               973 |
|          4 |          4 | 52.6 |    - |                 - |
|         10 |         10 | 57.6 |    - |                 - |
|         10 |       1000 | 71.1 |    - |                 - |
|       1000 |      20000 |  237 |    - |                 - |

Second, different numbers of nodes and edges, with 2 node types and 2 edge types (so 8 type triples, most probably): (NB. edges column first, because that makes the ordering/changes clearest)

| edges | nodes |  new |  old |
|------:|------:|-----:|-----:|
| 20000 |  1000 | 45.0 |  610 |
| 20000 | 10000 | 47.1 | 1270 |
| 40000 |  1000 | 92.7 | 2200 |
| 40000 | 10000 | 95.3 | 2410 |
| 80000 |  1000 |  183 |    - |
| 80000 | 10000 |  201 |    - |
| 80000 | 80000 |  199 |    - |


I think there's room for improvement too: here's the top few functions in a profile of the new stellargraph for the first row of the second table above:

```
       71    2.148    0.030    4.180    0.059 graph.py:675(create_graph_schema)
       73    0.780    0.011    0.780    0.011 {method 'argsort' of 'numpy.ndarray' objects}
  2841647    0.589    0.000    0.589    0.000 {built-in method __new__ of type object at 0x10f207fe0}
  5680012    0.559    0.000    0.559    0.000 {method 'add' of 'set' objects}
  2840568    0.532    0.000    1.120    0.000 <string>:12(__new__)
  1420142    0.282    0.000    0.282    0.000 graph.py:220(is_directed)
       71    0.079    0.001    0.862    0.012 arraysetops.py:297(_unique1d)
      571    0.031    0.000    0.064    0.000 base.py:4262(__getitem__)
```

### Code to produce this

<details>

```python
import pandas as pd
import numpy as np
import stellargraph as sg
import networkx as nx

def test(n, ntypes, e, etypes, sample=None, skip_networkx=False):
    nodes = {f"n_{x}": pd.DataFrame([], index=range(x, n, ntypes)) for x in range(ntypes)}
    edges = {f"e_{x}": pd.DataFrame(np.random.randint(0, n, size=(len(range(x, e, etypes)), 2)), columns=["source", "target"], index=range(x, e, etypes)) for x in range(etypes) }
    graph = sg.StellarGraph(nodes=nodes, edges=edges)
    info = graph.info()
    print(info if len(info) < 1000 else info[:1000] + "\n... (trimmed)")

    %timeit -n 10 graph.info()
    if skip_networkx:
        return

    nx_g = nx.from_pandas_edgelist(
        pd.concat([df.assign(label=x) for x, df in edges.items()]),
        edge_attr="label",
        create_using=nx.MultiGraph
    )
    for x, data in nx_g.nodes(data=True):
        data["label"] = f"n_{x % ntypes}"
    graph = sg.StellarGraph(nx_g)
    %timeit -n 10 graph.info(sample=sample)

test(1000, 1, 20000, 1)

test(1000, 1, 20000, 1, sample=20)

test(1000, 1, 20000, 2)

test(1000, 1, 20000, 2, sample=20)

test(1000, 2, 20000, 1)

test(1000, 2, 20000, 1, sample=20)

test(1000, 2, 20000, 2)

test(1000, 2, 20000, 2, sample=20)

test(1000, 4, 20000, 4, skip_networkx=True)

test(1000, 10, 20000, 10, skip_networkx=True)

test(1000, 10, 20000, 100, skip_networkx=True)

test(1000, 10, 20000, 1000, skip_networkx=True)

test(1000, 1000, 20000, 20000, skip_networkx=True)

test(1000, 2, 40000, 2)

test(1000, 2, 80000, 2, skip_networkx=True)

test(10000, 2, 20000, 2)

test(10000, 2, 40000, 2)

test(10000, 2, 80000, 2, skip_networkx=True)

test(80000, 2, 80000, 2, skip_networkx=True)

%prun test(1000, 2, 20000, 2, skip_networkx=True)
```

</details>

Reply:
I've removed it for now.